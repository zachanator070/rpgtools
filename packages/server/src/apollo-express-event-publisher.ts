import { EventPublisher } from "./types";
import { PubSub } from "apollo-server-express";
import { injectable } from "inversify";

@injectable()
export class ApolloExpressEventPublisher implements EventPublisher {
	pubsub = new PubSub();

	asyncIterator(events: string[]): AsyncIterator<any> {
		return this.pubsub.asyncIterator(events);
	}

	publish(event: string, payload: any): Promise<void> {
		return this.pubsub.publish(event, payload);
	}
}
