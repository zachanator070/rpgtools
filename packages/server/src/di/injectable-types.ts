import {DatabaseContext} from "../dal/database-context";

export const INJECTABLE_TYPES = {

	DomainEntity: Symbol.for("DomainEntity"),

	// domain entities
	Article: Symbol.for("Article"),
	Chunk: Symbol.for("Chunk"),
	File: Symbol.for("File"),
	Game: Symbol.for("Game"),
	Image: Symbol.for("Image"),
	Item: Symbol.for("Item"),
	Model: Symbol.for("Model"),
	Monster: Symbol.for("Monster"),
	PermissionAssignment: Symbol.for("PermissionAssignment"),
	Person: Symbol.for("Person"),
	Pin: Symbol.for("Pin"),
	Place: Symbol.for("Place"),
	Role: Symbol.for("Role"),
	ServerConfig: Symbol.for("ServerConfig"),
	User: Symbol.for("User"),
	WikiFolder: Symbol.for("WikiFolder"),
	World: Symbol.for("World"),
	Event: Symbol.for("Event"),
	Calendar: Symbol.for("Calendar"),

	// entity factories
	ArticleFactory: Symbol.for("ArticleFactory"),
	AclFactory: Symbol.for('AclFactory'),
	ChunkFactory: Symbol.for("ChunkFactory"),
	FileFactory: Symbol.for("FileFactory"),
	GameFactory: Symbol.for("GameFactory"),
	ImageFactory: Symbol.for("ImageFactory"),
	ItemFactory: Symbol.for("ItemFactory"),
	ModelFactory: Symbol.for("ModelFactory"),
	MonsterFactory: Symbol.for("MonsterFactory"),
	PersonFactory: Symbol.for("PersonFactory"),
	PinFactory: Symbol.for("PinFactory"),
	PlaceFactory: Symbol.for("PlaceFactory"),
	RoleFactory: Symbol.for("RoleFactory"),
	ServerConfigFactory: Symbol.for("ServerConfigFactory"),
	UserFactory: Symbol.for("UserFactory"),
	WikiFolderFactory: Symbol.for("WikiFolderFactory"),
	WikiPageFactory: Symbol.for("WikiPageFactory"),
	WorldFactory: Symbol.for("WorldFactory"),
	EventWikiFactory: Symbol.for("EventFactory"),
	CalendarFactory: Symbol.for("CalendarFactory"),

	CharacterFactory: Symbol.for('CharacterFactory'),
	CharacterAttributeFactory: Symbol.for('CharacterAttributeFactory'),
	PathNodeFactory: Symbol.for('NodePathFactory'),
	StrokeFactory: Symbol.for('StrokeFactory'),
	FogStrokeFactory: Symbol.for('FogStrokeFactory'),
	MessageFactory: Symbol.for('MessageFactory'),
	InGameModelFactory: Symbol.for('InGameModelFactory'),

	AgeFactory: Symbol.for("AgeFactory"),
	MonthFactory: Symbol.for("MonthFactory"),
	DayOfTheWeekFactory: Symbol.for("DayOfTheWeekFactory"),

	// repositories
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	FileRepository: Symbol.for("FileRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	EventWikiRepository: Symbol.for("EventWikiRepository"),
	CalendarRepository: Symbol.for("CalendarRepository"),
	StrokeRepository: Symbol.for('StrokeRepository'),
	FogStrokeRepository: Symbol.for('FogStrokeRepository'),

	SqlPermissionControlledRepository: Symbol.for('SqlPermissionControlledRepository'),

	// authorization policies
	ArticleAuthorizationPolicy: Symbol.for("ArticleAuthorizationPolicy"),
	ChunkAuthorizationPolicy: Symbol.for("ChunkAuthorizationPolicy"),
	FileAuthorizationPolicy: Symbol.for("FileAuthorizationPolicy"),
	GameAuthorizationPolicy: Symbol.for("GameAuthorizationPolicy"),
	ImageAuthorizationPolicy: Symbol.for("ImageAuthorizationPolicy"),
	ItemAuthorizationPolicy: Symbol.for("ItemAuthorizationPolicy"),
	ModelAuthorizationPolicy: Symbol.for("ModelAuthorizationPolicy"),
	MonsterAuthorizationPolicy: Symbol.for("MonsterAuthorizationPolicy"),
	PersonAuthorizationPolicy: Symbol.for("PersonAuthorizationPolicy"),
	PinAuthorizationPolicy: Symbol.for("PinAuthorizationPolicy"),
	PlaceAuthorizationPolicy: Symbol.for("PlaceAuthorizationPolicy"),
	RoleAuthorizationPolicy: Symbol.for("RoleAuthorizationPolicy"),
	ServerConfigAuthorizationPolicy: Symbol.for("ServerConfigAuthorizationPolicy"),
	UserAuthorizationPolicy: Symbol.for("UserAuthorizationPolicy"),
	WikiFolderAuthorizationPolicy: Symbol.for("WikiFolderAuthorizationPolicy"),
	WikiPageAuthorizationPolicy: Symbol.for("WikiPageAuthorizationPolicy"),
	WorldAuthorizationPolicy: Symbol.for("WorldAuthorizationPolicy"),
	CalendarAuthorizationPolicy: Symbol.for("CalendarAuthorizationPolicy"),
	FogStrokeAuthorizationPolicy: Symbol.for('FogStrokeAuthorizationPolicy'),
	StrokeAuthorizationPolicy: Symbol.for('StrokeAuthorizationPolicy'),

	// archive repositories
	ArchiveArticleRepository: Symbol.for("ArchiveArticleRepository"),
	ArchiveChunkRepository: Symbol.for("ArchiveChunkRepository"),
	ArchiveGameRepository: Symbol.for("ArchiveGameRepository"),
	ArchiveImageRepository: Symbol.for("ArchiveImageRepository"),
	ArchiveItemRepository: Symbol.for("ArchiveItemRepository"),
	ArchiveModelRepository: Symbol.for("ArchiveModelRepository"),
	ArchiveMonsterRepository: Symbol.for("ArchiveMonsterRepository"),
	ArchivePermissionAssignmentRepository: Symbol.for("ArchivePermissionAssignmentRepository"),
	ArchivePersonRepository: Symbol.for("ArchivePersonRepository"),
	ArchivePinRepository: Symbol.for("ArchivePinRepository"),
	ArchivePlaceRepository: Symbol.for("ArchivePlaceRepository"),
	ArchiveRoleRepository: Symbol.for("ArchiveRoleRepository"),
	ArchiveServerConfigRepository: Symbol.for("ArchiveServerConfigRepository"),
	ArchiveUserRepository: Symbol.for("ArchiveUserRepository"),
	ArchiveWikiFolderRepository: Symbol.for("ArchiveWikiFolderRepository"),
	ArchiveWikiPageRepository: Symbol.for("ArchiveWikiPageRepository"),
	ArchiveWorldRepository: Symbol.for("ArchiveWorldRepository"),
	ArchiveFileRepository: Symbol.for("ArchiveFileRepository"),

	// services
	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	RoleService: Symbol.for("RoleService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	// session contexts
	SessionContextFactory: Symbol.for("SessionContextFactory"),
	SecurityContextFactory: Symbol.for("SecurityContextFactory"),

	// 5e
	Open5eApiClient: Symbol.for("Open5eApiClient"),
	Dnd5eApiClient: Symbol.for("Dnd5eApiClient"),

	// delta
	DeltaFactory: Symbol.for("DeltaFactory"),

	// cache
	Cache: Symbol.for("Cache"),

	// archive
	ArchiveFactory: Symbol.for("ArchiveFactory"),

	ZipArchiveFactory: Symbol.for("ZipArchiveFactory"),
	ZipArchive: Symbol.for('ZipArchive'),

	// event publisher
	EventPublisher: Symbol.for("EventPublisher"),

	// server
	ApiServer: Symbol.for("ApiServer"),
	ServerProperties: Symbol.for("ServerProperties"),
	RpgToolsServer: Symbol.for('RpgToolsServer'),
	DbEngine: Symbol.for("DbEngine"),
	SqlDbEngine: Symbol.for('SqlDbEngine'),

	// data loaders
	ArticleDataLoader: Symbol.for("ArticleDataLoader"),
	ChunkDataLoader: Symbol.for("ChunkDataLoader"),
	FileDataLoader: Symbol.for("FileDataLoader"),
	GameDataLoader: Symbol.for("GameDataLoader"),
	ImageDataLoader: Symbol.for("ImageDataLoader"),
	ItemDataLoader: Symbol.for("ItemDataLoader"),
	ModelDataLoader: Symbol.for("ModelDataLoader"),
	MonsterDataLoader: Symbol.for("MonsterDataLoader"),
	PermissionAssignmentDataLoader: Symbol.for("PermissionAssignmentDataLoader"),
	PersonDataLoader: Symbol.for("PersonDataLoader"),
	PinDataLoader: Symbol.for("PinDataLoader"),
	PlaceDataLoader: Symbol.for("PlaceDataLoader"),
	RoleDataLoader: Symbol.for("RoleDataLoader"),
	ServerConfigDataLoader: Symbol.for("ServerConfigDataLoader"),
	UserDataLoader: Symbol.for("UserDataLoader"),
	WikiFolderDataLoader: Symbol.for("WikiFolderDataLoader"),
	WikiPageDataLoader: Symbol.for("WikiPageDataLoader"),
	WorldDataLoader: Symbol.for("WorldDataLoader"),
	CalendarDataLoader: Symbol.for('CalendarDataLoader'),
	EventDataLoader: Symbol.for('EventDataLoader'),

	// seeders
	RoleSeeder: Symbol.for("RoleSeeder"),
	ServerConfigSeeder: Symbol.for("ServerConfigSeeder"),

	// database context
	DatabaseContext: Symbol.for("DatabaseContext"),
	DatabaseContextFactory: Symbol.for("DatabaseContextFactory"),

	// repo mapper
	EntityMapper: Symbol.for("EntityMapper"),

	// mongodb
	FilterFactory: Symbol.for("FilterFactory"),
	MongoDbMigrationV40: Symbol.for("MongoDbMigrationV40"),


};
