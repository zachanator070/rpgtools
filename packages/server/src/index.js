import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
