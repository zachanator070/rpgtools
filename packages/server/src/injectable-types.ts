export const INJECTABLE_TYPES = {
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PermissionAssignmentRepository: Symbol.for("PermissionAssignmentRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	FileRepository: Symbol.for("FileRepository"),

	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	SessionContextFactory: Symbol.for("SessionContextFactory"),
	Open5eApiClient: Symbol.for("Open5eApiClient"),

	Cache: Symbol.for("Cache"),
};
