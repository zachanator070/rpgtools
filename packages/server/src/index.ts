import {container} from "./di/inversify";
import {INJECTABLE_TYPES} from "./di/injectable-types";
import RpgToolsServer from "./server/rpgtools-server";

const server = container.get<RpgToolsServer>(INJECTABLE_TYPES.RpgToolsServer);

(async () => {
	await server.start();
})();
