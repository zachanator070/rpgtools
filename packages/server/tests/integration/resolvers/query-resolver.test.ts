import { defaultTestingContextFactory } from "../DefaultTestingContextFactory";
import { container } from "../../../src/di/inversify";
import { INJECTABLE_TYPES } from "../../../src/di/injectable-types";
import {WikiFolderService, WorldService} from "../../../src/types";
import { ANON_USERNAME } from "@rpgtools/common/src/permission-constants";
import {
	GET_CURRENT_USER,
	GET_WIKI,
	GET_CURRENT_WORLD,
	GET_WORLDS, SEARCH_ROLES,
	SEARCH_USERS,
	WIKIS_IN_FOLDER
} from "@rpgtools/common/src/gql-queries";

process.env.TEST_SUITE = "query-resolver-test";

describe("query resolver", () => {
	const worldService = container.get<WorldService>(INJECTABLE_TYPES.WorldService);
	const wikiFolderService = container.get<WikiFolderService>(INJECTABLE_TYPES.WikiFolderService);
	let {
		server,
		mockSessionContextFactory,
		otherUser,
		otherUserSecurityContext,
		world,
		testRole,
		currentUser,
		testerSecurityContext,
		newFolder,
		otherPage,
		...testingContext
	} = defaultTestingContextFactory();

	describe("with world", () => {
		beforeEach(async () => {
			({
				mockSessionContextFactory,
				otherUser,
				otherUserSecurityContext,
				world,
				testRole,
				currentUser,
				testerSecurityContext,
				newFolder,
				otherPage,
			} = await testingContext.reset());
			mockSessionContextFactory.resetCurrentUser();
		});

		test("no current user", async () => {
			const result = await server.executeGraphQLQuery({ query: GET_CURRENT_USER });
			expect(result).toMatchSnapshot({
				data: {
					currentUser: {
						username: ANON_USERNAME,
					},
				},
				errors: undefined,
			});
		});

		test("users", async () => {
			const result = await server.executeGraphQLQuery({
				query: SEARCH_USERS,
				variables: { username: "tester2" },
			});
			expect(result).toMatchSnapshot({
				data: {
					users: {
						docs: expect.arrayContaining([
							expect.objectContaining({
								_id: expect.any(String),
							}),
						]),
					},
				},
				errors: undefined,
			});
		});

		test("world", async () => {
			const result = await server.executeGraphQLQuery({
				query: GET_CURRENT_WORLD,
				variables: { worldId: world._id.toString() },
			});
			expect(result).toMatchSnapshot();
		});

		test("worlds with one private and one public world", async () => {
			await worldService.createWorld("Azeroth", true, testerSecurityContext);
			const result = await server.executeGraphQLQuery({
				query: GET_WORLDS,
				variables: { page: 1 },
			});
			expect(result).toMatchSnapshot({
				data: {
					worlds: {
						docs: expect.arrayContaining([
							expect.objectContaining({
								_id: expect.any(String),
								wikiPage: expect.objectContaining({
									_id: expect.any(String),
								}),
							}),
						]),
					},
				},
				errors: undefined,
			});
		});

		test("wiki no permission", async () => {
			const result = await server.executeGraphQLQuery({
				query: GET_WIKI,
				variables: { wikiId: world.wikiPage },
			});
			expect(result).toMatchSnapshot({
				errors: expect.arrayContaining([
					expect.objectContaining({
						message: expect.any(String),
					}),
				]),
			});
		});

		describe("with authenticated user", () => {
			beforeEach(async () => {
				mockSessionContextFactory.setCurrentUser(currentUser);
			});

			test("current user", async () => {
				const result = await server.executeGraphQLQuery({ query: GET_CURRENT_USER });
				expect(result).toMatchSnapshot({
					data: {
						currentUser: {
							_id: expect.any(String),
							roles: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
								}),
							]),
						},
					},
					errors: undefined,
				});
			});

			test("world", async () => {
				const result = await server.executeGraphQLQuery({
					query: GET_CURRENT_WORLD,
					variables: { worldId: world._id.toString() },
				});
				expect(result).toMatchSnapshot({
					data: {
						world: {
							_id: expect.any(String),
							roles: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
									members: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
											username: expect.any(String),
										}),
									]),
									permissions: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
											subject: expect.objectContaining({
												_id: expect.any(String),
											}),
										}),
									]),
								}),
							]),
							rootFolder: {
								_id: expect.any(String),
								children: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
									}),
								]),
							},
							wikiPage: {
								_id: expect.any(String),
							},
							accessControlList: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
									subject: expect.objectContaining({
										_id: expect.any(String),
									}),
									roles: expect.arrayContaining([
										expect.objectContaining({
											_id: expect.any(String),
										}),
									]),
								}),
							]),
							pins: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
									map: expect.objectContaining({
										_id: expect.any(String),
									}),
									page: expect.objectContaining({
										_id: expect.any(String),
									}),
								}),
							]),
						},
					},
					errors: undefined,
				});
			});

			test("worlds with one private and one public world", async () => {
				await worldService.createWorld("Azeroth", false, testerSecurityContext);
				const result = await server.executeGraphQLQuery({
					query: GET_WORLDS,
					variables: { page: 1 },
				});
				expect(result).toMatchSnapshot({
					data: {
						worlds: {
							docs: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String),
									wikiPage: expect.objectContaining({
										_id: expect.any(String),
									}),
								}),
							]),
						},
					},
					errors: undefined,
				});
			});

			test("wiki", async () => {
				const result = await server.executeGraphQLQuery({
					query: GET_WIKI,
					variables: { wikiId: world.wikiPage },
				});
				expect(result).toMatchSnapshot({
					data: {
						wiki: {
							_id: expect.any(String),
							world: {
								_id: expect.any(String),
							},
							folder: {
								_id: expect.any(String),
							},
						},
					},
					errors: undefined,
				});
			});

			test("wikis in folder", async () => {
				const placesFolder = (await wikiFolderService.getFolders(testerSecurityContext, world._id, "Places", undefined)).filter(folder => folder.name === "Places");
				const result = await server.executeGraphQLQuery({
					query: WIKIS_IN_FOLDER,
					variables: { folderId: placesFolder[0]._id},
				});
				expect(result).toMatchSnapshot({
					data: {
						wikisInFolder: {
							docs: expect.arrayContaining([
								expect.objectContaining({
									_id: expect.any(String)
								})
							]),
							nextPage: null
						}
					},
					errors: undefined
				})
			});

			it("get roles", async () => {
				const result = await server.executeGraphQLQuery({
					query: SEARCH_ROLES,
					variables: {
						worldId: world._id,
					},
				});
				expect(result).toMatchSnapshot();
			});

			it("get roles permission denied", async () => {
				mockSessionContextFactory.resetCurrentUser();
				const result = await server.executeGraphQLQuery({
					query: SEARCH_ROLES,
					variables: {
						worldId: world._id,
					},
				});
				expect(result).toMatchSnapshot();
			});
		});
	});
});
