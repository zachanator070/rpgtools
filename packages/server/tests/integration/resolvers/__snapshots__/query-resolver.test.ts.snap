// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query resolver with world no current user 1`] = `
{
  "data": {
    "currentUser": {
      "_id": Any<String>,
      "currentWorld": null,
      "email": null,
      "roles": [],
      "username": "Anonymous",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world users 1`] = `
{
  "data": {
    "users": {
      "docs": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
        },
      ],
      "page": 1,
      "totalPages": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world wiki no permission 1`] = `
{
  "data": {
    "wiki": null,
  },
  "errors": ArrayContaining [
    ObjectContaining {
      "message": Any<String>,
    },
  ],
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user current user 1`] = `
{
  "data": {
    "currentUser": {
      "_id": Any<String>,
      "currentWorld": null,
      "email": "tester@gmail.com",
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
        },
      ],
      "username": "tester",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user get roles 1`] = `
{
  "data": {
    "roles": {
      "docs": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "accessControlList": Any<Array>,
          "members": Any<Array>,
          "world": {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user get roles permission denied 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: You do not have permission to read this World],
  ],
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user wiki 1`] = `
{
  "data": {
    "wiki": {
      "_id": Any<String>,
      "accessControlList": [],
      "canAdmin": true,
      "canWrite": true,
      "content": null,
      "coverImage": null,
      "folder": {
        "_id": Any<String>,
      },
      "mapImage": null,
      "name": "Earth",
      "pixelsPerFoot": 0,
      "type": "Place",
      "world": {
        "_id": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user wikis in folder 1`] = `
{
  "data": {
    "wikisInFolder": {
      "docs": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
        },
      ],
      "nextPage": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user with hosted game get a game 1`] = `
{
  "data": {
    "game": ObjectContaining {
      "_id": Any<String>,
      "characters": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "name": "tester",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user with hosted game get my games 1`] = `
{
  "data": {
    "myGames": ArrayContaining [
      ObjectContaining {
        "_id": Any<String>,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user world 1`] = `
{
  "data": {
    "world": {
      "_id": Any<String>,
      "accessControlList": ArrayContaining [
        ObjectContaining {
          "permission": Any<String>,
          "principal": ObjectContaining {
            "_id": Any<String>,
            "name": Any<String>,
          },
          "principalType": Any<String>,
        },
      ],
      "canAddModels": true,
      "canAddRoles": true,
      "canAdmin": true,
      "canHostGame": true,
      "canWrite": true,
      "name": "Earth",
      "rootFolder": {
        "_id": Any<String>,
        "accessControlList": [],
        "canAdmin": true,
        "canWrite": true,
        "children": ArrayContaining [
          ObjectContaining {
            "_id": Any<String>,
          },
        ],
        "name": "Earth",
      },
      "wikiPage": {
        "_id": Any<String>,
        "mapImage": null,
        "name": "Earth",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with authenticated user worlds with one private and one public world 1`] = `
{
  "data": {
    "worlds": ObjectContaining {
      "docs": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "wikiPage": ObjectContaining {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world with hosted game get a game permission denied 1`] = `
{
  "data": null,
  "errors": Any<Array>,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world world 1`] = `
{
  "data": {
    "world": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`query resolver with world worlds with one private and one public world 1`] = `
{
  "data": {
    "worlds": ObjectContaining {
      "docs": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "wikiPage": ObjectContaining {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
